spec_macros/src/derive_spec/spec_enum.rs

```diff
@@ -0,0 +1,524 @@
+//                    ##     #####     #########      ##                    //
+//                    ##     ##                ##     ##                    //
+//                    ##     #######   #########      ##                    //
+//////////////////////////////////////////////////////////////////////////////
+
+#[cfg(test)]
+mod tests {
+
+    use proc_macro2::TokenStream;
+    use quote::{format_ident, quote};
+    use syn::{Visibility, parse_quote};
+
+    use super::derive_spec_enum;
+
+    #[test]
+    fn test_derive_spec_enum_simple() {
+        let variants: syn::punctuated::Punctuated<syn::Variant, syn::Token![,]> = parse_quote! {
+            VariantA,
+            VariantB(u32),
+            VariantC(String, bool),
+        };
+
+        let orig_name = format_ident!("TestEnum");
+        let vis: Visibility = parse_quote! { pub };
+        let type_attrs: Vec<TokenStream> = vec![quote! { #[derive(Debug)] }];
+
+        let derived =
+            derive_spec_enum(variants, orig_name.clone(), vis.clone(), type_attrs.clone()).unwrap();
+
+        let expected_spec_enum = quote! {
+            #(#type_attrs )*
+            #vis enum TestEnumSpec {
+                VariantA,
+                VariantB(u32),
+                VariantC(String, bool),
+            }
+        };
+
+        let expected_try_from_impl = quote! {
+            impl std::convert::TryFrom<TestEnum> for TestEnumSpec {
+                type Error = String;
+
+                fn try_from(orig: TestEnum) -> Result<Self, Self::Error> {
+                    Ok(match orig {
+                        TestEnum::VariantA => TestEnumSpec::VariantA,
+                        TestEnum::VariantB( field_0 ) => TestEnumSpec::VariantB( field_0 ),
+                        TestEnum::VariantC( field_0, field_1 ) => TestEnumSpec::VariantC( field_0, field_1 )
+                    })
+                }
+            }
+        };
+
+        let expected_from_impl = quote! {
+            impl From<TestEnumSpec> for TestEnum {
+                fn from(original: TestEnumSpec) -> Self {
+                    match original {
+                        TestEnumSpec::VariantA => TestEnum::VariantA,
+                        TestEnumSpec::VariantB( field_0 ) => TestEnum::VariantB( field_0 ),
+                        TestEnumSpec::VariantC( field_0, field_1 ) => TestEnum::VariantC( field_0, field_1 )
+                    }
+                }
+            }
+        };
+
+        assert_eq!(derived.obj.to_string(), expected_spec_enum.to_string());
+        assert_eq!(
+            derived.try_from_impl.to_string(),
+            expected_try_from_impl.to_string()
+        );
+        assert_eq!(
+            derived.from_impl.to_string(),
+            expected_from_impl.to_string()
+        );
+    }
+
+    #[test]
+    fn test_derive_spec_enum_with_named_variant() {
+        let variants: syn::punctuated::Punctuated<syn::Variant, syn::Token![,]> = parse_quote! {
```

Comment:
Add the import on top: `use syn::punctuated::Punctuated;` and use the type directly here, and in the other places it applies to.
```suggestion
        let variants: Punctuated<syn::Variant, syn::Token![,]> = parse_quote! {
```

Reply:
Yes, this is a possible syntax, as the original one. I don't see the readability suffering, but we are loosing time in writing and fixing the code.

Fixed.